.github/workflows/env-teardown.yml

```diff
@@ -1,6 +1,6 @@
 name: Remove environment
 on:
-  pull_request:
+  pull_request_target:
```

Comment:
@peterstarling I don't think it will run the changed workflow file. It runs the workflow file in master - I was not able to trigger the pipeline without merging to master first (I tried it on my fork, see link)

Reply:
If users would be able to modify the workflow file in a fork the feature could not be used in any public repository at all (because it would allow to expose all secret variables). The documentation mentions that `pull_request_target` can e.g. be used to label PRs (if users could modify the workflow file, that would be insecure too)

Reply:
```Although the workflow runs in the context of the base of the pull request, you should make sure that you do not check out, build, or run untrusted code from the pull request with this event.```

We're checking out code, running it. Any code we're included from npm, will have full access to the repo and the secrets.

Reply:
Because of the if, we are only checking out and running code that originates from the `yldio/asap-hub` repository - that is a trusted origin.

Any code included from npm (by us) already has full access to the repo and secrets due to the other jobs.

Reply:
https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

```Be careful when using pull_request_target and only use it when you actually need the privileged context of the target repo available in your workflow, especially when combined with explicit handling of the contents of an untrusted PR.```

We don't need the privileged context of the target repo.

Reply:
@markcsims could you describe what you think a possible attack would look like?

Reply:
> Be careful when using pull_request_target and only use it when you actually need the privileged context of the target repo available in your workflow, especially when combined with explicit handling of the contents of an untrusted PR.

I'm pretty sure that the authors of this text did not think about the edge case that `pull_request` does not trigger when there are merge conflicts while `pull_request_target` does trigger. The text basically says: **Do not use this feature if another (unprivileged) feature can doe the same thing** - and that is definitely true.

The blog post describes what insecure workflows look like, see this example:
```
# INSECURE. Provided as an example only.
on:
  pull_request_target

jobs:
  build:
    name: Build and test
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
      with:
        ref: ${{ github.event.pull_request.head.sha }}

    - uses: actions/setup-node@v1
    - run: |
        npm install
        npm build

    - uses: completely/fakeaction@v2
      with:
        arg1: ${{ secrets.supersecret }}

    - uses: fakerepo/comment-on-pr@v1
      with:
        message: |
          Thank you!
```

The important part here, as the blog post describes, is that actions/checkout makes an **explicit** checkout of the pull request head. This will pull in code from the untrusted PR. The blog post also describes that a `actions/checkout` without an explicit ref is not a security problem - it will checkout the code from the master branch of `yldio/asap-hub`. 

See this code example:

```# The workflow is broken. DO NOT use it in production.
on: [push, pull_request_target]

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2
      - name: Build and test
        run: /bin/bash ./build.sh && /bin/bash ./runtests.sh
      - name: Report
        if: failure() && github.event.action != 'push'
        with: fancy/commenter@v1
          message: |
            Some checks have failed.
```

> The workflow is broken in the sense that it does not actually build the PR, but luckily is **not vulnerable**, since there is no explicit checkout of the actual PR contents or unsafe handling of those contents.

### Summary
- No untrusted code is being run
- Additionally there is a check that the PR originates from yldio/asap-hub
- I added **permissions** to set them to read-all so that no write access to the repo exists.

I'll start an implementation that removes PRs every day using cron 👍

Reply:
ok what you say makes a lot of sense @marot - to be fair I never really looked into this too closely but using this trigger would actually solve many of our problems.

As I understand we are unable to run a modified version of the workflow on a fork - however we are able to modify the code that this workflow runs. 

> If users would be able to modify the workflow file in a fork the feature could not be used in any public repository at all

To be fair I thought that any other workflow trigger does not run on a fork.

To prevent from checking out the code from the PR of the fork we added the `if`, correct? The part I don't fully understand is the explicit ref and how it could be insecure, but I take it that the `if` is good enough in this scenario.

I guess what we should do is test that we are really:
1. Unable to modify the workflow (i think you said you'd tested that already @marot ? can you post the proof?)
2. Unable to run the modified code of the PR in the fork. For this line for instance:
https://github.com/yldio/asap-hub/blob/d641bc80081b79042f86a83cf20a09a1eb3fd111/.github/workflows/env-teardown.yml#L48
We could rewrite the python script to access github secrets or something no? But as I understand this step will no longer run because of the `if`, still worth testing that. What I don't understand about the explicit `ref` is that I thought by default the checkout action without the explicit ref checks out to the pull request head, otherwise we would never be able to run the modified code in our PR workflows which we do all the time, right?

Reply:
The normal checkout action will checkout the code from the target repo (yldio/asap-hub). You can specify explicitly that you want to checkout code from the fork - we do not do that atm. Checking out code from a fork is safe if you do not execute any of the code but merely 'look' at it.

> What I don't understand about the explicit ref is that I thought by default the checkout action without the explicit ref checks out to the pull request head, otherwise we would never be able to run the modified code in our PR workflows which we do all the time, right?

Yeah, and I think the name of the trigger originates from that. `pull_request_**target**` -> it points to the target.

For testing, I think it is better to create a new repo on our personal accounts. It is better visible what is going on too

Reply:
Here is a test-workflow setup:
https://github.com/marot/test-workflows/runs/5568945338?check_suite_focus=true

![image](https://user-images.githubusercontent.com/74212402/158578351-c9f27361-b57e-461f-8b66-6db8b4087885.png) This is a job from the feature branch. You can see that the workflow is taken from master

In the script job, you can see the content from master (the branch has a modified message): 
![image](https://user-images.githubusercontent.com/74212402/158578499-4d99668d-4d79-4908-aa27-bbdd747af8a1.png)

Here you can see the job was skipped: https://github.com/marot/test-workflows/runs/5569041536?check_suite_focus=true

Here I disabled the if on master, and you can see that the script.sh from master is executed: https://github.com/marot/test-workflows/pull/2/checks

Reply:
One more thing, I would change the if to this: `if: ${{ github.event.inputs.prNumber || github.event.pull_request.head.repo.full_name == 'marot/test-workflows' }}` so that we can run the job also manually.
Only people with write access can trigger the workflow